cli/cook/subcommands/ssh.py

```diff
@@ -0,0 +1,112 @@
+import logging
+import os
+from operator import itemgetter
+from urllib.parse import urlparse
+
+from cook import colors, http
+from cook.subcommands.show import query, print_no_data
+from cook.util import print_info, strip_all
+
+
+def sandbox_directory(agent_work_dir, instance, job):
+    """
+    Constructs the agent sandbox directory given the agent working directory, an instance, and its parent job.
+    """
+    directory = os.path.join(agent_work_dir, 'slaves', instance["slave_id"], 'frameworks',
+                             job["framework_id"], 'executors', instance["executor_id"], 'runs', 'latest')
+    logging.debug(f'sandbox directory = {directory}')
+    return directory
+
+
+def ssh_to_instance(instance, job):
+    """Attempts to ssh (using os.execlp) to the Mesos agent corresponding to the given instance."""
+    hostname = instance['hostname']
+    if 'output_url' in instance:
+        output_url = instance['output_url']
+        url = urlparse(output_url)
+        netloc = url.netloc
+    else:
+        logging.info('assuming default agent port of 5051')
+        netloc = f'{hostname}:5051'
+
+    resp = http.__get(f'http://{netloc}/state')
+    if resp.status_code != 200:
+        logging.error(f'mesos agent returned status code {resp.status_code} and body {resp.text}')
+        raise Exception('Encountered error when querying Mesos agent for its work directory.')
+
+    resp_json = resp.json()
+    flags = resp_json['flags']
+    if 'work_dir' not in flags:
+        logging.error(f'work_dir was not present in mesos agent state flags: {flags}')
+        raise Exception('Mesos agent state did not include its work directory.')
+
+    print_info(f'Executing ssh to {colors.bold(hostname)}.')
+    agent_work_dir = flags['work_dir']
+    directory = sandbox_directory(agent_work_dir, instance, job)
+    file = os.environ.get('SSH', 'ssh')
+    logging.info(f'using ssh file: {file}')
+    os.execlp(file, 'ssh', '-t', hostname, f'cd {directory} ; bash')
+
+
+def ssh_to_job(job):
+    """
+    Tries ssh_to_instance for each instance from most to least recent, until either
+    ssh can be executed for an instance, or all instances have been tried and failed.
+    """
+    instances = job['instances']
+    if len(instances) == 0:
+        raise Exception('This job currently has no instances.')
+
+    for instance in sorted(instances, key=itemgetter('start_time'), reverse=True):
+        try:
+            print_info(f'Attempting ssh for job instance {colors.bold(instance["task_id"])}...')
+            ssh_to_instance(instance, job)
+        except Exception as e:
+            print_info(str(e))
+    else:
```

Comment:
Falling back to `ssh`ing into the non-latest instance might be confusion to users.

Reply:
Ditched this, and now we only try the most recently started instance